### PR TITLE
onetbb: add version 2021.6.0

### PR DESCRIPTION
**Pull request by kambala-decapitator**
Specify library name and version:  **onetbb/2021.6.0**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

Tested the following configurations with default options locally:

- 2021.6.0 macOS Intel (native)
- 2021.6.0 macOS arm64 (cross)
- 2021.6.0 iOS arm64 (cross)
- 2021.3.0 macOS Intel (native)

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by uilianries** [comment]
It seems like the test_package will need an update too.

**Comment by kambala-decapitator** [comment]
> It seems like the test_package will need an update too.

I've contacted author of duplicate PR #13116 to incorporate my additional changes in there. Or should we do it vice versa?

**Comment by uilianries** [comment]
@kambala-decapitator You can do vice-versa, but be careful with git conflicts.

**Comment by kambala-decapitator** [comment]
I don't think I understand why it fails even with the workaround applied... Locally on macOS it works correctly: the test executable doesn't run in this specific case.

File: recipes/onetbb/all/conanfile.py [hunk 1]
```diff
@@ -26,17 +27,21 @@ class OneTBBConan(ConanFile):
...
         "fPIC": True,
         "tbbmalloc": False,
         "tbbproxy": False,
+        "interprocedural_optimization": True,
     }
 
     def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
+        if not (Version(self.version) >= "2021.6.0" and self.options.shared and self.settings.os != "Android"):
```
[the remark this answers is not in the document]
Comment (kambala-decapitator) [reply recipes/onetbb/all/conanfile.py:43]: I took this condition from onetbb itself: https://github.com/oneapi-src/oneTBB/blob/c8504c4ec866eb474d323e9df326778f30f9850d/CMakeLists.txt#L187-L190

would be great if you could fix it in upstream

[the remark this answers is not in the document]
Comment (cguentherTUChemnitz) [reply recipes/onetbb/all/conanfile.py:43]: Ok i will try to test the upstream project for this. My insights so far were, that this is heavily dependent on version of the ndk. Since the last years the Android NDK was going to a full gcc to clang migration. I am not sure how this should be handled correctly, since it might fail in some ndk versions. But i think it is fine to go for now with your working state and an dedicated search on how to enable IPO for Android later on. It should not block the effort for now.

TLDR: fine for me now ;)

File: recipes/onetbb/all/test_package/conanfile.py [hunk 2]
```diff
@@ -21,5 +26,9 @@ def build(self):
 
     def test(self):
         if not cross_building(self):
+            if TestPackageConan.should_skip_executing:
+                self.output.warn("Not running the built test executable for debug+static build to avoid assertion error, "
+                    "see https://github.com/oneapi-src/oneTBB/issues/920 for more info")
```
Comment (prince-chrismc) [inline recipes/onetbb/all/test_package/conanfile.py:31]: Should this not be blocked at the recipe level if it does not work?

File: recipes/onetbb/all/test_package/conanfile.py [hunk 3]
```diff
@@ -7,10 +7,15 @@
 class TestPackageConan(ConanFile):
     settings = "os", "arch", "compiler", "build_type"
     generators = "CMakeToolchain", "CMakeDeps", "VirtualRunEnv"
+    should_skip_executing = False
```
Comment (prince-chrismc) [inline recipes/onetbb/all/test_package/conanfile.py:10]: this would be a first in CCI so I am not sure this is the best way since consumer would have the same problem... 

Why publish a package that we know does not work?

**Comment by kambala-decapitator** [comment]
@prince-chrismc @uilianries would it work if I replace the test program with one of official examples from https://github.com/oneapi-src/oneTBB/tree/master/examples? For example, I see they have a Fibonacci example at https://github.com/oneapi-src/oneTBB/tree/master/examples/test_all/fibonacci

**Comment by uilianries** [comment]
@kambala-decapitator We don't care much about are what code is used, only if it's short and uses the generated package, and links to its libraries. Said that, Fibonacci is too long, 500 lines. Please, simplify it in case you want to change the test package.

Still the most important is that verify if the package is working, test package only validates those generated artifacts.

**Comment by kambala-decapitator** [comment]
OneTBB themselves don't support static builds: https://github.com/oneapi-src/oneTBB/blob/13544f9432c946d11cbdd39b4726407c8aa9e3a9/CMakeLists.txt#L116

should the recipe disable static configuration then? Or at least issue a warning?

File: recipes/onetbb/all/test_package/conanfile.py [hunk 4]
```diff
@@ -21,5 +26,9 @@ def build(self):
 
     def test(self):
         if not cross_building(self):
```
Comment (jwillikers) [inline recipes/onetbb/all/test_package/conanfile.py:28]: I recommend using `can_run` here.

File: recipes/onetbb/all/test_package/conanfile.py (hunk 3 above)
Comment (jwillikers) [reply recipes/onetbb/all/test_package/conanfile.py:10]: Yeah, I think it's important this be tested as part of the test package. TBB underpins STL Parallel Algorithms in GCC, unless that's changed since I used it. I think it's critical that we verify that the runtime is integrated properly.

**Review by jwillikers — CHANGES_REQUESTED** [review]

**Review by prince-chrismc — COMMENTED** [review]
I would put a validate  heck and fail instead of hiding it in the test package.

Put a message like "this doesn't seem to work... please open a pr if you can fix .."

We allow not building every configuration as long as it's not a regression :)

**Comment by kambala-decapitator** [comment]
> I would put a validate heck and fail instead of hiding it in the test package.
> 
> Put a message like "this doesn't seem to work... please open a pr if you can fix .."
> 
> We allow not building every configuration as long as it's not a regression :)

thanks, did like that

**Review by uilianries — COMMENTED** [review]
LGTM. Waiting for @prince-chrismc review too.

**Review by prince-chrismc — APPROVED** [review]
LGTM

**Review by jwillikers — APPROVED** [review]
